// likec4/__app__/src/chunks/-index-overview-fhfhEsGb.js
import { jsx as t, jsxs as u, Fragment as z } from "react/jsx-runtime";
import { useOverviewGraph as L } from "virtual:likec4/overview-graph";
import { nonexhaustive as R } from "@likec4/core";
import { u as V } from "./tanstack-router-B97XgOdm.js";
import { B as G, H as v, P as g, u as $, a as K, i as U, b as A, c as X } from "./likec4-DgZMMrvs.js";
import { memo as k, useRef as Y, useMemo as B } from "react";
import { c as b, I as q, n as S, u as j, i as J } from "./main-DiEdSc7f.js";
import { P as O, c as I, G as C, T as F, a as y, C as Q, b as W, d as ee, I as te, B as P, u as se } from "./mantine-Dy50iVbn.js";
import { usePreviewUrl as ae } from "virtual:likec4/previews";
/**
 * @license @tabler/icons-react v3.29.0 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
var oe = b("outline", "loader", "IconLoader", [["path", { d: "M12 6l0 -3", key: "svg-0" }], ["path", { d: "M16.25 7.75l2.15 -2.15", key: "svg-1" }], ["path", { d: "M18 12l3 0", key: "svg-2" }], ["path", { d: "M16.25 16.25l2.15 2.15", key: "svg-3" }], ["path", { d: "M12 18l0 3", key: "svg-4" }], ["path", { d: "M7.75 16.25l-2.15 2.15", key: "svg-5" }], ["path", { d: "M6 12l-3 0", key: "svg-6" }], ["path", { d: "M7.75 7.75l-2.15 -2.15", key: "svg-7" }]]);
/**
 * @license @tabler/icons-react v3.29.0 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
var re = b("filled", "file-filled", "IconFileFilled", [["path", { d: "M12 2l.117 .007a1 1 0 0 1 .876 .876l.007 .117v4l.005 .15a2 2 0 0 0 1.838 1.844l.157 .006h4l.117 .007a1 1 0 0 1 .876 .876l.007 .117v9a3 3 0 0 1 -2.824 2.995l-.176 .005h-10a3 3 0 0 1 -2.995 -2.824l-.005 -.176v-14a3 3 0 0 1 2.824 -2.995l.176 -.005h5z", key: "svg-0" }], ["path", { d: "M19 7h-4l-.001 -4.001z", key: "svg-1" }]]), ie = "mxt2a80";
function ne(r) {
  return r.reduce((e, [s, a], o) => e + `${o === 0 ? "M" : " L"} ${s},${a}`, "");
}
function de({
  id: r,
  data: e,
  ...s
}) {
  if (!e)
    return null;
  const a = ne(e.points);
  return /* @__PURE__ */ t(
    G,
    {
      id: r,
      path: a,
      ...s
    }
  );
}
var h = "_1n8mzjc0", _ = "_1n8mzjc1", D = "_1n8mzjc2", M = "_1n8mzjc3", le = "_1n8mzjc4", ce = "_1n8mzjc5", me = "_1n8mzjc6", ue = "_1n8mzjc7", pe = "_1n8mzjc8";
const fe = /* @__PURE__ */ k(function({
  data: e,
  parentId: s,
  id: a
}) {
  const o = S(s);
  return /* @__PURE__ */ u(z, { children: [
    /* @__PURE__ */ t(v, { type: "target", position: g.Top, className: h }),
    /* @__PURE__ */ t(
      O,
      {
        p: "sm",
        pt: "xs",
        radius: "md",
        withBorder: !0,
        shadow: o ? "lg" : "xs",
        className: I(
          le,
          o ? _ : D,
          e.dimmed && M
        ),
        children: /* @__PURE__ */ u(C, { gap: 8, children: [
          /* @__PURE__ */ t(F, { size: 24, variant: "transparent", color: "dark.4", children: /* @__PURE__ */ t(q, { size: "100%" }) }),
          /* @__PURE__ */ t(y, { size: "lg", fw: 500, children: e.label })
        ] })
      }
    ),
    /* @__PURE__ */ t(v, { type: "source", position: g.Bottom, className: h })
  ] });
}), ve = /* @__PURE__ */ k(function({
  data: e,
  parentId: s,
  id: a
}) {
  const o = S(s);
  return /* @__PURE__ */ u(z, { children: [
    /* @__PURE__ */ t(v, { type: "target", position: g.Top, className: h }),
    /* @__PURE__ */ t(
      O,
      {
        p: "sm",
        pt: "xs",
        radius: "md",
        withBorder: !0,
        shadow: o ? "lg" : "xs",
        className: I(
          ce,
          o ? _ : D,
          e.dimmed && M
        ),
        children: /* @__PURE__ */ u(C, { gap: 8, children: [
          /* @__PURE__ */ t(F, { size: 24, variant: "transparent", color: "dark.3", children: /* @__PURE__ */ t(re, { size: "100%" }) }),
          /* @__PURE__ */ t(y, { size: "lg", fw: 500, children: e.label })
        ] })
      }
    ),
    /* @__PURE__ */ t(v, { type: "source", position: g.Bottom, className: h })
  ] });
}), ge = /* @__PURE__ */ k(function({
  data: e,
  height: s = 320
}) {
  const o = ae(e.viewId);
  return /* @__PURE__ */ u(z, { children: [
    /* @__PURE__ */ t(v, { type: "target", position: g.Top, className: h }),
    /* @__PURE__ */ u(
      Q,
      {
        className: I(
          me,
          e.dimmed && M
        ),
        withBorder: !0,
        shadow: "xs",
        padding: 0,
        children: [
          /* @__PURE__ */ t(W, { className: ue, children: o ? /* @__PURE__ */ t(
            te,
            {
              src: o,
              fit: "contain",
              h: s - 60
            }
          ) : /* @__PURE__ */ t(ee, { h: s - 60, children: /* @__PURE__ */ u(C, { children: [
            /* @__PURE__ */ t(F, { size: 60, variant: "transparent", color: "dark", children: /* @__PURE__ */ t(oe, { stroke: 1.5, size: "100%" }) }),
            /* @__PURE__ */ t(y, { size: "xl", fw: 500, c: "dimmed", children: "Preview not available" })
          ] }) }) }),
          /* @__PURE__ */ t(P, { className: pe, h: 60, p: "sm", pl: "md", children: /* @__PURE__ */ t(y, { component: "div", size: "lg", fw: 500, children: e.label }) })
        ]
      }
    ),
    /* @__PURE__ */ t(v, { type: "source", position: g.Bottom, className: h })
  ] });
}), he = {
  folder: fe,
  file: ve,
  view: ge
}, we = {
  link: de
}, ye = (r) => ({
  nodes: r.nodes.map(({ id: e, parentId: s, position: a, width: o, height: c, ...l }) => {
    const i = s ? r.nodes.find((N) => N.id === s) : null, w = {
      ...a,
      width: o,
      height: c
    };
    i && (a = {
      x: a.x - i.position.x,
      y: a.y - i.position.y
    });
    const p = i ? { parentId: i.id } : {};
    switch (l.type) {
      case "file":
      case "folder":
        return {
          id: e,
          type: l.type,
          data: {
            dimmed: !1,
            label: l.label,
            path: l.path,
            rect: w
          },
          deletable: !1,
          position: a,
          width: o,
          height: c,
          zIndex: 1,
          ...p
        };
      case "view":
        return {
          id: e,
          type: "view",
          data: {
            dimmed: !1,
            label: l.label,
            viewId: l.viewId,
            rect: w
          },
          selectable: !1,
          deletable: !1,
          position: a,
          width: o,
          height: c,
          zIndex: 3,
          ...p
        };
      default:
        R(l);
    }
  }),
  edges: r.edges.map((e) => ({
    id: e.id,
    source: e.source,
    target: e.target,
    type: "link",
    zIndex: 2,
    hidden: !0,
    data: {
      points: e.points
    }
  }))
});
function Ne({
  graph: r,
  fitViewPadding: e = 0.1,
  zoomable: s = !0,
  pannable: a = !0
}) {
  const o = V(), c = Y(void 0), { colorScheme: l } = se(), i = B(() => ye(r), [r]), [w, p, N] = $(i.nodes), [E, H, Z] = K(i.edges);
  j(() => {
    p(
      (n) => i.nodes.map((d) => {
        const f = n.find((m) => m.id === d.id);
        return f ? { ...J(f, ["selected", "hidden"]), ...d } : d;
      })
    ), H(i.edges);
  }, [i.nodes, i.edges]);
  const x = w.find((n) => n.selected);
  return j(() => {
    const n = c.current;
    n && (x ? n.fitView({
      maxZoom: 1,
      padding: e,
      nodes: [x],
      duration: 450
    }) : n.fitView({
      maxZoom: 1,
      padding: e,
      duration: 800
    }));
  }, [x?.id ?? null]), /* @__PURE__ */ t(
    U,
    {
      colorMode: l === "auto" ? "system" : l,
      className: ie,
      nodeTypes: he,
      edgeTypes: we,
      nodes: w,
      onNodesChange: N,
      edges: E,
      onEdgesChange: Z,
      fitView: !0,
      fitViewOptions: B(() => ({
        minZoom: 0.05,
        maxZoom: 1,
        padding: e,
        includeHiddenNodes: !0
      }), [e]),
      nodesDraggable: !1,
      nodesConnectable: !1,
      nodesFocusable: !0,
      edgesReconnectable: !1,
      edgesFocusable: !1,
      multiSelectionKeyCode: null,
      zoomOnPinch: s,
      zoomOnScroll: !a && s,
      zoomOnDoubleClick: !1,
      ...!s && {
        zoomActivationKeyCode: null
      },
      maxZoom: s ? 2 : 1,
      minZoom: s ? 0.01 : 1,
      preventScrolling: s || a,
      noDragClassName: "nodrag",
      noPanClassName: "nopan",
      panOnScroll: a,
      panOnDrag: a,
      ...!a && {
        selectionKeyCode: null
      },
      onInit: (n) => c.current = n,
      onNodeClick: (n, d) => {
        if (d.type === "view") {
          n.stopPropagation(), p((f) => f.map(({ data: m, ...T }) => ({ ...T, data: { ...m, dimmed: T.id !== d.id } }))), c.current?.fitView({
            maxZoom: 10,
            padding: 0,
            nodes: [d],
            duration: 1200
          }), setTimeout(() => {
            c.current?.updateNodeData(d.id, { dimmed: !0 });
          }, 400), setTimeout(() => {
            o.navigate({
              to: "/view/$viewId/",
              params: {
                viewId: d.data.viewId
              },
              search: !0
            });
          }, 800);
          return;
        }
        d.selected && (n.stopPropagation(), p((f) => f.map((m) => m.id === d.id ? { ...m, selected: !1 } : m)));
      },
      children: /* @__PURE__ */ t(A, { variant: X.Dots, size: 4, gap: 50 })
    }
  );
}
function je() {
  const r = L();
  return /* @__PURE__ */ t(P, { pos: "fixed", inset: 0, children: /* @__PURE__ */ t(Ne, { graph: r }) });
}
export {
  je as default
};
